File: backend/internal/services/bookmark_service.go
package services

import (
	"backend/internal/models"
	"backend/internal/repositories/interfaces"
)

type BookmarkService struct {
	repo interfaces.BookmarkRepositoryInterface
}

func NewBookmarkService(repo interfaces.BookmarkRepositoryInterface) *BookmarkService {
	return &BookmarkService{repo: repo}
}

func (s *BookmarkService) GetBookmarkedNovelsByUserID(userID uint, page, limit int) ([]models.Novel, int64, error) {
	return s.repo.GetBookmarkedNovelsByUserID(userID, page, limit)
}

func (s *BookmarkService) GetBookmarkByUserIDAndNovelID(userID uint, novelID string) (models.BookmarkedNovel, error) {
	return s.repo.GetBookmarkByUserIDAndNovelID(userID, novelID)
}

func (s *BookmarkService) UpdateBookmark(novel models.BookmarkedNovel) (models.BookmarkedNovel, error) {
	return s.repo.UpdateBookmark(novel)
}

func (s *BookmarkService) CreateBookmark(bookmarkedNovel models.BookmarkedNovel) (*models.BookmarkedNovel, error) {
	return s.repo.CreateBookmark(bookmarkedNovel)
}

func (s *BookmarkService) UnbookmarkNovel(userID uint, novelID uint) error {
	return s.repo.DeleteBookmark(userID, novelID)
}

File: frontend/interfaces/services/BookmarkService.ts
import type { BookmarkedNovel, BookmarkedNovelSchema } from '~/schemas/BookmarkedNovel';
import type { PaginatedServerResponse } from '~/schemas/PaginatedServerResponse';
import type { SuccessServerResponse } from '~/schemas/SuccessServerResponse';
import type { NovelSchema } from '~/schemas/Novel';

export interface BookmarkService {
    bookmarkNovel(novelId: number, userId: number): Promise<BookmarkedNovel>;
    updateBookmark(updatedBookmark: BookmarkedNovel): Promise<BookmarkedNovel>;
    unbookmarkNovel(novelId: number, userId: number): Promise<SuccessServerResponse>;
    fetchBookmarkedNovelByUser(novelId: string, userId: number): Promise<BookmarkedNovel>;
    fetchBookmarkedNovelsByUser(userId: number, page: number, limit: number): Promise<PaginatedServerResponse<typeof NovelSchema>>;
}
